pkg/services/secretstore/service.go

```diff
@@ -0,0 +1,106 @@
+package secretstore
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/grafana/pkg/bus"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/secrets"
+	"github.com/grafana/grafana/pkg/services/sqlstore"
+)
+
+type Service struct {
+	Bus            bus.Bus
+	SQLStore       *sqlstore.SQLStore
+	SecretsService secrets.Service
+
+	scDecryptionCache secureJSONDecryptionCache
+}
+
+type secureJSONDecryptionCache struct {
+	cache map[int64]cachedDecryptedJSON
+	sync.Mutex
+}
+
+type cachedDecryptedJSON struct {
+	updated time.Time
+	json    map[string]string
+}
+
+func ProvideService(bus bus.Bus, store *sqlstore.SQLStore, secretsService secrets.Service) *Service {
+	s := &Service{
+		Bus:            bus,
+		SQLStore:       store,
+		SecretsService: secretsService,
+		scDecryptionCache: secureJSONDecryptionCache{
+			cache: make(map[int64]cachedDecryptedJSON),
+		},
+	}
+
+	s.Bus.AddHandler(s.GetSecrets)
+	s.Bus.AddHandler(s.GetSecret)
+	s.Bus.AddHandler(s.AddSecret)
+	s.Bus.AddHandler(s.DeleteSecret)
+	s.Bus.AddHandler(s.UpdateSecret)
+
+	return s
+}
+
+type SecretRetriever interface {
+	GetSecret(ctx context.Context, query *models.GetSecretQuery) error
+}
+
+func (s *Service) GetSecret(ctx context.Context, query *models.GetSecretQuery) error {
```

Comment:
Just thinking out loud here and obviously aware that big refactoring and changes are needed to support below and/or be a future idea. I'm not even sure it makes sense from a product perspective. In addition, note that I'm biased towards Kubernetes way of defining/handling secrets. 

Feels like a simpler interface would be exposing a (single) secret by key, rather than by orgId/uid, would maybe be more in line with how these external secret managers work/expose their interfaces? For example [getting a secret using Vault](https://learn.hashicorp.com/tutorials/vault/getting-started-first-secret#read-a-secret). . 

Data sources would hold reference to a secret in the `jsonData` for each secret used. identifying them in the secrets store by key, e.g. `key=org/<orgId>/<secrets name>`. 

Lets say that the Prometheus datasource expose possibility to configure basic authentication, e.g. via this [shared component](https://github.com/grafana/grafana/blob/e814e7364bebe80c61c2439e9f54f908a7dcce87/packages/grafana-ui/src/components/DataSourceSettings/BasicAuthSettings.tsx#L38-L60). 

`jsonData`:
```json
{
  "basicAuthUser: "user1"
}
```
`secureJsonData`:
```json
{
  "basicAuthPassword: "pwd2"
}
```

With the approach I have in mind you would store a reference to the basic auth secret:
HTTP POST /api/secrets (orgId=1)
```json
  "key": "org/1/prometheus-basic-auth-secret",
  "type": "basic auth",
  "value": {
    "username": "user1",
    "password": "pwd2"
  }
```

Then for a Prometheus datasource you would store a reference to that in `jsonData`:
```json
{
  "basicAuth": {
    "secretRef": "org/1/prometheus-basic-auth-secret"
  }
}
```

Example listing all configured secrets in an org::
HTTP GET /api/secrets (orgId=1)
In turn calling this secrets store, `store.GetSecrets(ctx, "org/1/").

In general quite similar to our [k/v store](https://github.com/grafana/grafana/blob/e814e7364bebe80c61c2439e9f54f908a7dcce87/pkg/infra/kvstore/kvstore.go#L23).